### PR TITLE
RFC: Guidelines for Accepting Component-level Contributions

### PR DESCRIPTION
**Pull request by sclevine**
[Readable](https://github.com/buildpacks/rfcs/blob/component-level-contrib/text/0000-component-level-contrib.md).

### CODE REVIEW

File: text/0000-component-level-contrib.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+
+# Motivation
+[motivation]: #motivation
+
+Our current contribution guidelines do not define a process for accepting component-level contributios, such as existing git repositories, that were developed outside of the project.
```
Comment (jkutner) [inline text/0000-component-level-contrib.md:15]:
```suggestion
Our current contribution guidelines do not define a process for accepting component-level contributions, such as existing git repositories, that were developed outside of the project.
```

File: text/0000-component-level-contrib.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+- Accepted guidelines may be too rigid and exclude components that the project should accept
+- Accepted guidelines may not be rigid enough, and the project may feel pressured to accept components that are inappropriate
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Do not establish guidelines for component-level contributions
+
+
```
Comment (jkutner) [inline text/0000-component-level-contrib.md:51]: Is there any prior art? Surely we're not the first open source project to confront this.

File: text/0000-component-level-contrib.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+- Supersedes: N/A
+
+
+# Motivation
+[motivation]: #motivation
+
+Our current contribution guidelines do not define a process for accepting component-level contributios, such as existing git repositories, that were developed outside of the project.
+While the lack of guidelines has not prevented components from being contributed (e.g., #123), establishing formal guidelines will ensure fair and orderly consideration of future proposals.
+
+# Proposed Guidelines
```
Comment (jkutner) [inline text/0000-component-level-contrib.md:18]:
```suggestion
# What it is
[what-it-is]: #what-it-is
```

File: text/0000-component-level-contrib.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
+# Meta
+[meta]: #meta
+- Name: Guidelines for Component-level Contributions
+- Start Date: 2021-03-08
+- Author(s): @sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
```
[the remark this answers is not in the document]
Comment (sclevine) [reply text/0000-component-level-contrib.md:10]: Thanks, sounds good. Will commit this later with the "in our the organization" typo fixed

File: text/0000-component-level-contrib.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+# Motivation
+[motivation]: #motivation
+
+Our current contribution guidelines do not define a process for accepting component-level contributions, such as existing git repositories, that were developed outside of the project.
+While the lack of guidelines has not prevented components from being contributed (e.g., #123), establishing formal guidelines will ensure fair and orderly consideration of future proposals.
```
Comment (ekcasey) [inline text/0000-component-level-contrib.md:16]:
```suggestion
While the lack of guidelines has not prevented components from being contributed (e.g., [#123](https://github.com/buildpacks/rfcs/pull/123)), establishing formal guidelines will ensure fair and orderly consideration of future proposals.
```
Special github linking for things like #123 don't appear to work in the body of a markdown document.

File: text/0000-component-level-contrib.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+- If a proposed contribution provides functionality that is already provided by the project, such functionality must be provided for use in a non-overlapping context, or the proposal must include a plan for consolidation.
+  - Qualifying examples:
+    1. Buildpack language bindings for a language other than Go
+    2. An alternative implementation of the buildpack lifecycle launcher in Rust, with a documented plan to deprecate the original Go version
+  - Non-qualifying examples:
+    1. An alternative implementation of the lifecycle in Rust that would exist concurrently with the Go version in perpetuity
+- If a proposed contribution is an adapter or integration with another service or technology, that technology must be notable and/or widely used (at the Core Team’s discretion). 
+- If a component is proposed for donation by one or more Core Team members individually, those Core Team members must abstain from voting over its acceptance.
+- If a component is proposed for donation by one or more Core Team members’ employer or employers, those Core Team members must abstain from voting over its acceptance.
```
Comment (ekcasey) [inline text/0000-component-level-contrib.md:38]: Have we clarified anywhere how abstentions affect the voting math. I assume we need a supermajority of non-abstaining voters?

**Review by ekcasey — APPROVED** [review]

File: text/0000-component-level-contrib.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+# What it is
+[what-it-is]: #what-it-is
+[proposed-guidelines]: #proposed-guidelines
+
+- Component-level contributions must be proposed as project-wide RFCs
+- Proposed contributions must primarily support the goals and use cases of the project and may not contain unrelated functionality that would not otherwise be accepted via the RFC process.
+  - Qualifying examples:
+    1. A Github Action that performs buildpack builds
+    2. A Concourse resource that creates builder images
```
Comment (hone) [inline text/0000-component-level-contrib.md:26]:
```suggestion
    1. A Concourse resource that creates builder images
```

File: text/0000-component-level-contrib.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    3. A v2 Heroku buildpack that runs Cloud Native Buildpack (v3) builds on platforms that only support older v2 buildpacks.
```
Comment (hone) [inline text/0000-component-level-contrib.md:27]:
```suggestion
    1. A v2 Heroku buildpack that runs Cloud Native Buildpack (v3) builds on platforms that only support older v2 buildpacks.
```

File: text/0000-component-level-contrib.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+  - Non-qualifying examples:
+    1. A generic CI/CD tool that includes functionality to run buildpack builds, but primarily supports use cases outside of the project (e.g., Tekton)
+- If a proposed contribution provides functionality that is already provided by the project, such functionality must be provided for use in a non-overlapping context, or the proposal must include a plan for consolidation.
+  - Qualifying examples:
+    1. Buildpack language bindings for a language other than Go
+    2. An alternative implementation of the buildpack lifecycle launcher in Rust, with a documented plan to deprecate the original Go version
```
Comment (hone) [inline text/0000-component-level-contrib.md:33]:
```suggestion
    1. An alternative implementation of the buildpack lifecycle launcher in Rust, with a documented plan to deprecate the original Go version
```

File: text/0000-component-level-contrib.md (hunk 2 above)
Comment (hone) [reply text/0000-component-level-contrib.md:52]: In a previous WG, there was an ask to look at what the CNCF polity is here b/c there may be legal implications with transferring IP etc. @sclevine have you had a chance to investigate?

**Review by nebhale — APPROVED** [review]

File: text/0000-component-level-contrib.md (hunk 2 above)
Comment (sclevine) [reply text/0000-component-level-contrib.md:57]: Clarified that these contributions are still subject to CNCF/LF review/approval. Is that good enough for this RFC?

File: text/0000-component-level-contrib.md (hunk 6 above)
Comment (sclevine) [reply text/0000-component-level-contrib.md]: Added wording to clarify

File: text/0000-component-level-contrib.md (hunk 2 above)
Comment (hone) [reply text/0000-component-level-contrib.md:57]: We should clarify who is responsible doing this work. I'll make a suggestion.

File: text/0000-component-level-contrib.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    1. Buildpack language bindings for a language other than those already supported by the project (such as Go)
+    1. An alternative implementation of the buildpack lifecycle launcher in Rust, with a documented plan to deprecate the original Go version
+  - Non-qualifying examples:
+    1. An alternative implementation of the lifecycle in Rust that would exist concurrently with the Go version in perpetuity
+- If a proposed contribution is an adapter or integration with another service or technology, that technology must be notable and/or widely used (at the Core Team’s discretion). 
+- If a component is proposed for donation by one or more Core Team members individually, those Core Team members must abstain from voting over its acceptance.
+- If a component is proposed for donation by one or more Core Team members’ employer or employers, those Core Team members must abstain from voting over its acceptance. Supermajority consensus of non-abstaining Core Team members is still required.
+- Component-level contributions are subject legal review by the Cloud Native Computing Foundation / Linux Foundation. Adherance to these guidances does not guarantee acceptance.
```
Comment (hone) [inline text/0000-component-level-contrib.md:43]:
```suggestion
- Component-level contributions are subject legal review by the Cloud Native Computing Foundation / Linux Foundation. Adherance to these guidances does not guarantee acceptance. The investigation of this is to be done by the Core Team before acceptance.
```

**Review by hone — APPROVED** [review]

**Review by jkutner — APPROVED** [review]

**Comment by sclevine** [comment]
FCP as of  5/12

**Comment by sclevine** [comment]
@hone to create issue for community repo language, and/or @sclevine to PR copy of text

**Comment by samj1912** [comment]
When this is added to buildpacks/community we may want to consider this comment from @dwillist https://github.com/buildpacks/rfcs/pull/159#discussion_r632038799

**Pull request merged**